### ISSUE TITLE
LLVM crashes when performing peephole optimisation on functions

### ISSUE THREAD

**Issue by MunchDev** [llvm/llvm-project#53297]
Context:
- MacOS 12.1
- LLVM installed via `brew`
- Output of `llvm-config --version`: `13.0.0`

C++ sample code is `fib(n)`:
```
int a(b) {
    if (b < 3) return 1;
    else return a(b-1)+a(b-2);
}
```

IR printed by calling `func->print(llvm::outs())` (without optimisation):
```
define double @a(double %b) {
entry:
  %cmp_lt_tmp = fcmp ule double %b, 3.000000e+00
  %bool_tmp = uitofp i1 %cmp_lt_tmp to double
  %if_cond = fcmp one double %bool_tmp, 0.000000e+00
  br i1 %if_cond, label %then, label %else

then:                                             ; preds = %entry
  br label %if_cont

else:                                             ; preds = %entry
  %sub_tmp = fsub double %b, 1.000000e+00
  %0 = call double @a(double %sub_tmp)
  %sub_tmp1 = fsub double %b, 2.000000e+00
  %1 = call double @a(double %sub_tmp1)
  %add_tmp = fadd double %0, %1
  br label %if_cont

if_cont:                                          ; preds = %else, %then
  %if_tmp = phi double [ 1.000000e+00, %then ], [ %add_tmp, %else ]
  ret double %if_tmp
}
```
(I'm writing a custom language so the type is `double`, but it should still crashes if it were an `i`)

When I use `llvm::legacy::FunctionPassManager` and add `llvm::createInstructionCombiningPass()`, the program crashes here:
```cpp
/* ... */
llvm::legacy::FunctionPassManager manager;
/* ... */
manager.add(llvm::createInstructionCombiningPass());
Function *func;
/* ... */
manager.run(*func); // Crashes here
```

The generate IR when compiled to bitcode (with `llvm-as`) and optimised (with `opt -O3`) does not have any issues, so errors in code generation are `likely` ruled out.

Any ideas?

**Comment by zero9178** [comment]
Given the IR is handled correctly by `opt`, and other LLVM tooling, it sounds more likely that your code is somehow interfacing with LLVM APIs incorrectly which leads to crashes of some sorts. As to why a version switch to LLVM 13 from LLVM 11 would result in that can be speculated. 
If there previously was Undefined Behaviour there isn't really anything to reason about. What could have possibly happened as well is that an API change somewhere between LLVM 11 and LLVM 13 results in issues in your code, which would be very surprising.

Have you tried running a verifier over the `llvm::Function` before running the optimizer? Another thing to try would be linking against an LLVM build with asserts if that is possible.

**Comment by MunchDev** [comment]
Yep, I did in fact run a verifier before attempting optimisation.
But here's the full code involving optimisation (for more context):
```
class CodeGen {
    llvm::LLVMContext context;
    llvm::IRBuilder<> builder;
    llvm::Module module;
    llvm::legacy::FunctionPassManager passManager;

    CodeGen() : context(), builder(context), module("Main", context), passManager(module) {
        passManager.add(llvm::createInstructionCombiningPass());
        passManager.add(llvm::createReassociatePass());
        passManager.add(llvm::createGVNPass());
        passManager.add(llvm::createCFGSimplificationPass());
    }
}

/* ... */

// Create return value
Builder().CreateRet(body_code);

llvm::verifyFunction(*func);
passManager.run(*func);
```

**Comment by MunchDev** [comment]
Hmm, I noticed that I didn't call `passManager.doInitialization();`, so it wasn't really a bug. Not sure what it initialises, but it isn't needed for many other passes other than `llvm::createInstructionCombiningPass()`.

**Comment by fhahn** [comment]
> Hmm, I noticed that I didn't call passManager.doInitialization();, so it wasn't really a bug. Not sure what it initialises, but it isn't needed for many other passes other than llvm::createInstructionCombiningPass().

From that I take it the issue has been resolved? I'll close it for now, please feel free to re-open if there is still an issue with LLVM
